File: events/2015/19/solution.go
package main

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"slices"
	"strings"
	"time"
	"yod_aoc/utils/go/aoc_utils"
)

func findSubstitutes(medicine string, molecule string, replacement string) []string {
	medLen := len(medicine)
	molLen := len(molecule)

	if medLen < molLen {
		return []string{}
	}

	// newMedicine := ""
	foundMedicines := []string{}
	newMedicine := ""

	for i := range medLen - molLen + 1 {
		testPart := medicine[i : i+molLen]

		if molecule == testPart {
			new := newMedicine + replacement + medicine[i+molLen:]
			foundMedicines = append(foundMedicines, new)
		}
		newMedicine += medicine[i : i+1]
	}

	return foundMedicines
}

// ----------------------------------------------------------------------------------------------------
// | Part 1
// ----------------------------------------------------------------------------------------------------
func part1(input []string) int {
	replacements := make(map[string][]string)
	foundMedicines := []string{}
	medicine := ""

	for index := range input {
		line := input[index]

		if len(line) == 0 {
			continue
		}

		if strings.Contains(line, " => ") {
			parts := strings.Split(line, " => ")

			if replacements[parts[0]] == nil {
				replacements[parts[0]] = []string{parts[1]}
			} else {
				replacements[parts[0]] = append(replacements[parts[0]], parts[1])
			}
		} else {
			medicine = line
		}
	}

	for molecule, value := range replacements {
		for i := range value {
			substitute := value[i]

			results := findSubstitutes(medicine, molecule, substitute)

			for _, res := range results {
				if !slices.Contains(foundMedicines, res) {
					foundMedicines = append(foundMedicines, res)
				}
			}
		}
	}

	return len(foundMedicines)
}

// ----------------------------------------------------------------------------------------------------
// | Part 2
// ----------------------------------------------------------------------------------------------------
func reverse(input string) string {
	asRune := []rune(input)

	for i, j := 0, len(asRune)-1; i < j; i, j = i+1, j-1 {
		asRune[i], asRune[j] = asRune[j], asRune[i]
	}

	result := string(asRune)
	return result
}

func part2(input []string) int {
	replacements := make(map[string]string)
	medicine := ""

	for index := range input {
		line := input[index]

		if len(line) == 0 {
			continue
		}

		if strings.Contains(line, " => ") {
			parts := strings.Split(line, " => ")

			replacements[reverse(parts[1])] = reverse(parts[0])
		} else {
			medicine = reverse(line)
		}
	}

	steps := 0

	for len(medicine) > 1 {
		bestIndex := math.MaxInt
		bestFrom := ""
		bestTo := ""

		for from, to := range replacements {
			index := strings.Index(medicine, from)

			if index != -1 && index < bestIndex {
				bestIndex = index
				bestFrom = from
				bestTo = to
			}
		}

		if bestFrom != "" {
			medicine = medicine[:bestIndex] + bestTo + medicine[bestIndex+len(bestFrom):]
			steps += 1
		} else {
			fmt.Println(fmt.Errorf("Stuck after %d steps", steps))
			return -1
		}
	}

	return steps
}

// ----------------------------------------------------------------------------------------------------
// | Main Function
// ----------------------------------------------------------------------------------------------------
func main() {
	// Read the arguments
	args := os.Args[1:]
	var fileName string
	if len(args) > 0 {
		fileName = args[0]
	} else {
		fileName = "input.txt"
	}
	// Start the timer!
	startTime := time.Now()

	// Read the input to an array of strings
	path := aoc_utils.GetFilePath(2015, 19, fileName)
	content, err := aoc_utils.ReadInputAsArray(path)
	if err != nil {
		fmt.Println("Error reading input:", err)
		return
	}

	// Run the Parts
	p1Result := aoc_utils.MeasurePerformance(func() interface{} {
		return part1(content)
	})

	p2Result := aoc_utils.MeasurePerformance(func() interface{} {
		return part2(content)
	})

	// End the Timer!
	endTime := time.Now()
	duration := endTime.Sub(startTime)

	resultsForDay := aoc_utils.AOCDayResults{
		Year:     2015,
		Day:      19,
		Part1:    p1Result,
		Part2:    p2Result,
		Duration: duration.Nanoseconds(),
		Timestamp: aoc_utils.AOCTimestamp{
			Start: startTime.Format(time.RFC3339),
			End:   endTime.Format(time.RFC3339),
		},
	}
	jsonResults, _ := json.Marshal(resultsForDay)
	fmt.Println(string(jsonResults))
}

File: events/2015/19/solution.py
import sys
import json
import time
from datetime import timedelta, datetime
from pathlib import Path

# Add workspace root to path for imports
workspace_root = Path(__file__).parent.parent.parent.parent
sys.path.insert(0, str(workspace_root))

from utils.python.types import AOCDayResults, AOCTimestamp, AOCPartResult
from utils.python.input import readInputAsArray
from utils.python.performance import measurePerformance

def findSubstitutes(medicine: str, molecule: str, replacement: str) -> list[str]:
  medLen = len(medicine)
  molLen = len(molecule)

  if medLen < molLen:
    return []

  newMedicine = ""
  foundMedicines: list[str] = []

  for i in range(medLen-molLen):
    testPart = medicine[i:i+molLen]

    if testPart == molecule:
      foundMedicines.append(newMedicine + replacement + medicine[i+molLen:])
    
    newMedicine += medicine[i]

  return foundMedicines


# ----------------------------------------------------------------------------------------------------
# | Part 1
# ----------------------------------------------------------------------------------------------------
def part1(input: list[str]) -> int:
  replacements: dict[str, list[str]] = {}
  medicine: str = ""
  foundMedicines: list[str] = []

  for i in input:
    if len(i) == 1 or i == " ":
      continue

    if "=>" in i:
      parts = i.split(" => ")

      if parts[0] in replacements:
        replacements[parts[0]].append(parts[1].rstrip())
      else:
        replacements[parts[0]] = [parts[1].rstrip()]
    else:
      medicine = i

  for molecule in replacements:
    for substitute in replacements[molecule]:
      results = findSubstitutes(medicine, molecule, substitute)

      for res in results:
        if res not in foundMedicines:
          foundMedicines.append(res)

  return len(foundMedicines)

# ----------------------------------------------------------------------------------------------------
# | Part 2
# ----------------------------------------------------------------------------------------------------
def part2(input: list[str]) -> int:
  replacements: dict[str, str] = {}
  medicine: str = ""

  for i in input:
    if len(i) == 1 or i == " ":
      continue

    if "=>" in i:
      parts = i.split(" => ")

      replacements[parts[1].rstrip().strip()[::-1]] = parts[0].rstrip().strip()[::-1]
    else:
      medicine = i[::-1].rstrip().strip()

  steps = 0

  while medicine != "e":
    bestIndex = float('inf')
    bestFrom = ""
    bestTo = ""

    for fromMolecule, toMolecule in replacements.items():
      index = medicine.find(fromMolecule)
      if index != -1 and index < bestIndex:
        bestIndex = index
        bestFrom = fromMolecule
        bestTo = toMolecule

    if bestIndex != float('inf'):
      medicine = medicine[:bestIndex] + bestTo + medicine[bestIndex + len(bestFrom):]
      steps += 1
    else:
      raise Exception(f'Stuck after {steps} steps {medicine} {len(medicine)}')

  return steps

# ----------------------------------------------------------------------------------------------------
# | Main Function
# ----------------------------------------------------------------------------------------------------
def main():
  # Get the arguments
  args = sys.argv[1:]
  fileName = args[0] if args else "input.txt"

  # Start the timer
  start_timestamp = datetime.now().isoformat()
  start_time = time.monotonic()

  # Read the input to an array of strings
  input = readInputAsArray(fileName)

  ## Run the parts
  p1Result = measurePerformance(part1, input)
  p2Result = measurePerformance(part2, input)

  # End the timer
  end_time = time.monotonic()
  end_timestamp = datetime.now().isoformat()
  duration = timedelta(seconds=end_time - start_time).total_seconds()*1000000000 # Convert to nanoseconds

  # Return the results
  return AOCDayResults(
    Year=2015,
    Day=19,
    Part1=p1Result,
    Part2=p2Result,
    Duration=duration,
    Timestamp=AOCTimestamp(
      Start=start_timestamp,
      End=end_timestamp,
    ),
  )



if __name__ == "__main__":
  results = main()
  print(json.dumps(results))